apps/src/templates/progress/googlePlatformApiRedux.js

```diff
@@ -0,0 +1,139 @@
+import firehoseClient from '@cdo/apps/lib/util/firehose';
+
+/*
+ * The Google Classroom Share Button is only available through the google
+ * platform api, so we have to add it to our page on load. We manage loading
+ * in redux since multiple components need to observe the loading state.
+ */
+
+const START_LOADING_GAPI = 'googlePlatformApi/START_LOADING_GAPI';
+const FINISH_LOADING_GAPI = 'googlePlatformApi/FINISH_LOADING_GAPI';
+
+const startLoadingGapi = startTime => ({type: START_LOADING_GAPI, startTime});
+const finishLoadingGapi = success => ({type: FINISH_LOADING_GAPI, success});
+
+export const GOOGLE_PLATFORM_API_ID = 'GooglePlatformApiId';
+const LOAD_TIMEOUT_MILLIS = 15000;
+
+const initialState = {
+  loading: false,
+  loaded: false,
+  loadStartTime: null
+};
+
+export function loadGooglePlatformApi() {
+  return (dispatch, getState) => {
+    if (gapiReady()) {
+      dispatch(finishLoadingGapi(true));
+      return Promise.resolve();
+    } else {
+      dispatch(startLoadingGapi(Date.now()));
+      return new Promise((resolve, reject) => {
+        const promise = {resolve: resolve, reject: reject};
+        loadApi(dispatch, getState, promise);
```

Comment:
While we _can_ trigger ajax requests in redux, I would recommend against it. Redux is a global, predictable, state container. When it starts doing too much, it becomes difficult to reason about. It should only handle mutations in state. When an engineer opens a redux file, they should expect to see how various input map to various outputs. If we start having redux trigger ajax requests and multiple dispatches, redux stops being a state container, starts to turn into a global "does everything" manager.

Reply:
after all the reading i've done on async actions in redux, i have to disagree here. or rather, argue that this is just an async mutation of state.

first, my interpretation is that the redux file isn't about mapping various inputs to various outputs, but rather, mapping various _actions_ to various _states_ – which i think is exactly what is happening here.

the official redux docs have a whole [section on api fetching](https://redux.js.org/tutorials/essentials/part-5-async-logic), which implies they consider an async mutation of state that uses remotely-fetched data to update the local state to be a valid use of the redux store. in this case we're dispatching an action to make a request to load the api, which results in a mutation of state to either `loaded=true` or `failed=true` (with the state of `loading=true` in the interim).

all that said, what are you suggesting i should do here instead? just move all the api loading stuff into a different file? or some more fundamental change to the logic?

Reply:
(Commenting in reverse order) 
My suggestion would be to use this redux file for only the state portions and move the api loading into a "loader" module. The "loader" module would then be what dispatches the actions.

Reply:
Redux does offer support for api fetching, but I have yet to see it used elegantly. It almost by definition breaks the "mapping various actions to various states" - an action is triggered that starts a load process, but it could end in 1 of 3 different states (loading, success, and error)

Reply:
If you're up for it, I'd prefer to raise this question to the whole engineering org and allow multiple people to weigh in.

Reply:
re. first comment:
if this were a long complicated file i would totally agree, and i definitely did consider doing that initially, but in this case i would argue that separating the loading logic out would actually make all this harder to read because you'd need both files open side-by-side to make sense of it, and it would likely add a lot more lines of boilerplate code. so in this case where the file is less than 150 lines, i personally think it's easiest to understand having it all in one place.

another thing i like about keeping it in one file is that all of the state and logic is internal to this loading process, so right now we only export the reducer, `loadGooglePlatformApi`, and `canShowGoogleShareButton` (plus `GOOGLE_PLATFORM_API_ID` for testing), which are the only parts a consumer should care about. the only part of the state that can be subscribed to (without looking in this file or examining the state at runtime) is the `loaded` property, and even that is obscured by `canShowGoogleShareButton`. so to the extent that someone might view the list of exports as the "external api", i think this module is exposing exactly what it should. if we separate it out into a loader, everything would be exposed, giving it the potential to be misused.

additionaly, it turns out the redux docs actually [encourage](https://redux.js.org/style-guide/style-guide#structure-files-as-feature-folders-or-ducks) putting all the logic in a single file if you're not using the "feature folder" approach (which we most definitely are not). not to mention that in the context of our codebase i think this is the general expectation.

Reply:
re. second comment:
what is the alternative in an app that uses redux and requires async requests? _something_ has to track the progress of the request, and synchronizing some external object with the redux state doesn't sound elegant to me.

we know that there are three actions that need to update the state of this redux branch:

1) page load => start loading api :: state.loading = true
2) load fails => finish loading api (false) :: state.loading = false, state.loaded = false
3) load succeeds => finish loading api (success=true) :: state.loading = false, state.loaded = true

so we actually have three actions resulting in three different states (though in this case the failure case isn't made explicit, because our application doesn't need to know about it).

it is true that a single action initiates the transition from the initial state to one of three resulting states, but that is the only valid way to initiate that transition (there's no valid way to set `loaded=true` other than dispatching that first async action). and it does require a subsequent event (the api finishes loading or fails) to transition past the state caused by the first action.

also, i guess i'm not sure what this comment is suggesting i should change. even if i move the loading logic to a different module, it would still need to dispatch an async action, because _by definition_ the transition to `loading=true` needs to be atomically bound to the initiation of the loading process. do you have some other proposal for how our redux-based application should handle async requests?

Reply:
re. comment 3:
i personally don't think this is the time/place to initiate that conversation with the larger team unless we're at an impasse here. this feature is _way_ overdo and that's a big conversation to have as an org, so in my opinion the fact that we already do this a fair amount in our code means we're not introducing some new bad practice, just deferring the cleanup of an (arguably) existing one.

i think your first comment mostly amounts to an issue of style and code organization, and i don't have a strong enough opinion about it to delay this feature anymore, so if you feel strongly that this stuff should be separated into two files i will do that.

as for your second comment, i would suggest that you come up with a proposal for how we _should_ keep redux store in sync with the state of ajax requests before raising the issue with the broader team.

Reply:
Thank you for the context on timing/urgency for this feature. You are completely right - it's definitely _not_ worth delaying this feature to get this all figured out. I'm going to go ahead and retract this suggestion on this PR. This PR was a great catalyst to get this discussion started, but the discussion and decision should be separate from the PR.